diff --git a/CMakeLists.txt b/CMakeLists.txt
index 006e93a..c863f35 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -62,11 +62,17 @@ else()
     endif()
 endif()
 
+if (RT64_HOST_DXC)
+    set (DXC ${RT64_HOST_DXC})
+endif()
+
 set(ZSTD_LEGACY_SUPPORT OFF)
 set(ZSTD_BUILD_STATIC ON)
 set(ZSTD_BUILD_SHARED OFF)
 
-add_subdirectory(src/tools/file_to_c)
+if (NOT TARGET file_to_c)
+    add_subdirectory(src/tools/file_to_c)
+endif()
 add_subdirectory(src/contrib/re-spirv)
 add_subdirectory(src/contrib/nativefiledialog-extended)
 add_subdirectory(src/contrib/zstd/build/cmake)
@@ -444,12 +450,18 @@ target_link_libraries(rt64 libzstd_static)
 target_link_libraries(rt64 plume)
 
 # Add tools.
-add_subdirectory(src/tools/texture_hasher)
-add_subdirectory(src/tools/texture_packer)
+if (NOT TARGET texture_hasher)
+    add_subdirectory(src/tools/texture_hasher)
+endif()
+if (NOT TARGET texture_packer)
+    add_subdirectory(src/tools/texture_packer)
+endif()
 
 # Add any Apple-specific source files and libraries
 if (APPLE)
-    add_subdirectory(src/tools/spirv_cross_msl)
+    if (NOT TARGET spirv_cross_msl)
+        add_subdirectory(src/tools/spirv_cross_msl)
+    endif()
 
     target_sources(rt64 PRIVATE
         "${PROJECT_SOURCE_DIR}/src/apple/rt64_apple.mm"
